### PR TITLE
feat: Add num_partitions field in mongodbatlas_search_index

### PR DESCRIPTION
**Pull request by filipcirtog**
## Description

The numPartitions field for MongoDB Atlas Search Index exists in the API, but Terraform can't manage it yet. The API overwrites nested fields like definition.numPartitions entirely (similar to PUT), making partial updates via PATCH unsupported. For now, customers need to use the MongoDB Atlas UI to manage this field until Terraform adds support.

This PR is for the addition of the `num_partitions` field in `mongodbatlas_search_index`

Link to any related issue(s):  CLOUDP-365383

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [x] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR. A migration guide must be created or updated if the new feature will go in a major version.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR. A migration guide must be created or updated.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contributing guides](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/contributing/README.md)
- [x] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code
- [ ] If changes include deprecations or removals I have added appropriate changelog entries.
- [ ] If changes include removal or addition of 3rd party GitHub actions, I updated our internal document. Reach out to the APIx Integration slack channel to get access to the internal document.

### CODE REVIEW

File: docs/data-sources/search_index.md [hunk 1]
```diff
@@ -46,5 +46,6 @@ data "mongodbatlas_search_index" "test" {
 * `type_sets` - Set of type set definitions (when present). Each item includes:
   * `name` - Type set name.
   * `types` - JSON array string describing the types for the set.
+* `num_partitions` - (Optional) Number of index partitions. Allowed values are [1, 2, 4].
```
Comment (QuiqueSanMongo) [inline docs/data-sources/search_index.md]: Does the feature just allow 1, 2, and 4, or is this a typo?

**Review by QuiqueSanMongo — APPROVED** [review]
LGTM, just a small question

File: internal/service/searchindex/resource_search_index_test.go [hunk 2]
```diff
@@ -538,6 +541,43 @@ func configVector(projectID, indexName, clusterName string) string {
 	`, clusterName, projectID, indexName, database, collection, fieldsJSON)
 }
 
+func configSearchWithNumPartitions(projectID, indexName, clusterName string, numPartitions int) string {
+	return fmt.Sprintf(`
+
+		resource "mongodbatlas_search_deployment" "test" {
+			project_id   = %[2]q
```
Comment (marcabreracast) [inline internal/service/searchindex/resource_search_index_test.go:603]: [nit]: Change the order (`cluster_name` before `project_id`) so it matches the formatting options

**Comment by marcabreracast** [comment]
I see that the PR is marked as `Open`, but the `Required Checklist:` section of the PR description is not marked, nor the `Type of change` section. Since we have some automations that get triggered depending on the change, shouldn't this be considered as a **New Feature**? 
As we're adding a new field to the `mongodbatlas_search_index`, I'd expect that a new changelog will also be generated, that will be then included in the release notes (example: [PR #3978](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/3978)

**Comment by filipcirtog** [comment]
@marcabreracast thank you! I have converted it back to draft until I address your feedback!

File: docs/data-sources/search_index.md (hunk 1 above)
Comment (filipcirtog) [reply docs/data-sources/search_index.md]: thank you! yess, link to the api [definition in here](https://www.mongodb.com/docs/api/doc/atlas-admin-api-v2/operation/operation-creategroupclusterftsindex). additionally, I've updated the definition with the default value:
>  Number of index partitions. Allowed values are [1, 2, 4]. Default value is 1.

File: internal/service/searchindex/resource_search_index_test.go (hunk 2 above)
Comment (filipcirtog) [reply internal/service/searchindex/resource_search_index_test.go:603]: thanks! i have updated this!

**Comment by filipcirtog** [comment]
@marcabreracast Thank you. I have updated the PR `title`, `changelog` and `Required Checklist`.

File: docs/data-sources/search_index.md [hunk 3]
```diff
@@ -46,5 +46,6 @@ data "mongodbatlas_search_index" "test" {
 * `type_sets` - Set of type set definitions (when present). Each item includes:
   * `name` - Type set name.
   * `types` - JSON array string describing the types for the set.
+* `num_partitions` - Number of index partitions. Allowed values are [1, 2, 4]. Default value is 1.
```
Comment (lantoli) [inline docs/data-sources/search_index.md]: nit: for both datasource i think just "* `num_partitions` - Number of index partitions." is enough,  allowed values and default value is more insteresting for the resource

File: docs/resources/search_index.md [hunk 4]
```diff
@@ -237,6 +237,8 @@ EOF
     EOF
   ```
 
+* `num_partitions` - (Optional) Number of index partitions. Allowed values are [1, 2, 4]. Default value is 1.
```
Comment (lantoli) [inline docs/resources/search_index.md:240]: qq: can it be used in both search and vectorSearch types?

File: internal/service/searchindex/data_source_search_index.go [hunk 5]
```diff
@@ -118,6 +118,10 @@ func returnSearchIndexDSSchema() map[string]*schema.Schema {
 				},
 			},
 		},
+		"num_partitions": {
+			Type:     schema.TypeInt,
+			Computed: true,
```
Comment (lantoli) [inline internal/service/searchindex/data_source_search_index.go:123]: can you confirm that if num_partitions is not used when creating the resource, we'll get null in the data sources (instead of default 1)

**Review by QuiqueSanMongo — APPROVED** [review]
LGTM!

File: internal/service/searchindex/resource_search_index_test.go [hunk 6]
```diff
@@ -191,6 +191,9 @@ func TestAccSearchIndex_updatedToEmptyMappingsFields(t *testing.T) {
 func TestAccSearchIndex_withVector(t *testing.T) {
 	resource.ParallelTest(t, *basicVectorTestCase(t))
 }
+func TestAccSearchIndex_withNumPartitions(t *testing.T) {
+	resource.ParallelTest(t, *numPartitionsTestCase(t))
```
Comment (lantoli) [inline internal/service/searchindex/resource_search_index_test.go]: in this case that we're not reusing the test (e.g. for acceptance and migration test), you can inline `numPartitionsTestCase` and just have the code in TestAccSearchIndex_withNumPartitions

**Review by lantoli — APPROVED** [review]

File: internal/service/searchindex/resource_search_index_test.go [hunk 7]
```diff
@@ -546,6 +586,43 @@ func checkVector(projectID, indexName, clusterName string) resource.TestCheckFun
...
+		indexName              = acc.RandomName()
+	)
+	return &resource.TestCase{
+		PreCheck:                 func() { acc.PreCheckBasic(tb) },
+		ProtoV6ProviderFactories: acc.TestAccProviderV6Factories,
+		CheckDestroy:             acc.CheckDestroySearchIndex,
+		Steps: []resource.TestStep{
+			{
+				Config: configSearchWithNumPartitions(projectID, indexName, clusterName, 2),
+				Check:  checkSearchWithNumPartitions(projectID, indexName, clusterName, 2),
```
Comment (oarbusi) [inline internal/service/searchindex/resource_search_index_test.go]: similar to what Leo mentioned in another comment, it would be nice to test the case of when we don't set the num_partitions, that in the resource the field is not set/nil, and in the data source it has the default value 1 (i assume the API returns the value even if not explitictly set)

**Review by oarbusi — APPROVED** [review]
LGTM, just one test case suggestion

File: internal/service/searchindex/resource_search_index_test.go (hunk 2 above)
Comment (marcabreracast) [reply internal/service/searchindex/resource_search_index_test.go:603]: Thank you!

**Review by marcabreracast — APPROVED** [review]
LGTM, thank you for addressing my suggestion. Agreed with Leo and Oriol on the test suggestions, feel free to apply those.

File: .changelog/3982.txt [hunk 8]
```diff
@@ -0,0 +1,11 @@
+```release-note:enhancement
+resource/mongodbatlas_search_index: Adds the `num_partitions` attribute
```
Comment (AgustinBettati) [inline .changelog/3982.txt]:
```suggestion
resource/mongodbatlas_search_index: Adds `num_partitions` attribute 
```
nit: Just to stay consistent with other entries (applies to the 3 entries) https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/contributing/changelog-process.md#resource-and-provider-enhancements

**Review by AgustinBettati — APPROVED** [review]

File: internal/service/searchindex/resource_search_index_test.go (hunk 7 above)
Comment (AgustinBettati) [reply internal/service/searchindex/resource_search_index_test.go]: would also revise behaviour if attribute is initially defined and then removed

File: .changelog/3982.txt (hunk 8 above)
Comment (filipcirtog) [reply .changelog/3982.txt]: thank you! changed.

File: internal/service/searchindex/resource_search_index_test.go (hunk 6 above)
Comment (filipcirtog) [reply internal/service/searchindex/resource_search_index_test.go]: thank you! changed.

File: docs/resources/search_index.md (hunk 4 above)
Comment (filipcirtog) [reply docs/resources/search_index.md:240]: thank you! added the a test for vectorSearch as well.

File: docs/data-sources/search_index.md (hunk 3 above)
Comment (filipcirtog) [reply docs/data-sources/search_index.md]: thank you! updated!

File: internal/service/searchindex/resource_search_index_test.go [hunk 9]
```diff
@@ -533,11 +585,118 @@ func configVector(projectID, indexName, clusterName string) string {
...
+		checks = append(checks,
+			resource.TestCheckResourceAttr(resourceName, "num_partitions", fmt.Sprintf("%d", *numPartitions)),
+			resource.TestCheckResourceAttr(datasourceName, "num_partitions", fmt.Sprintf("%d", *numPartitions)),
+		)
+	} else {
+		var ret int
+		checks = append(checks,
+			resource.TestCheckResourceAttr(resourceName, "num_partitions", strconv.Itoa(ret)),
+			resource.TestCheckResourceAttr(datasourceName, "num_partitions", strconv.Itoa(ret)),
+		)
```
[the remark this answers is not in the document]
Comment (filipcirtog) [reply internal/service/searchindex/resource_search_index_test.go]: I tried to replicate the actual behavior of the resource/data source. Specifically, the situation where an integer pointer is received as nil. Since SDKv2 does not natively handle nil values, and our num_partitions field is of type integer, in such cases, the field will default to an undefined integer value. 

To simulate this behavior in Go, even when we receive a nil, it is processed as zero. Here’s the code snippet that demonstrates this.
```
if o == nil || IsNil(o.NumPartitions) 
{ var ret int return ret }
```

File: internal/service/searchindex/resource_search_index_test.go (hunk 7 above)
Comment (filipcirtog) [reply internal/service/searchindex/resource_search_index_test.go]: I have added some more tests as you, Leo and Agustin suggested. However, since we are not able to handle the returned nil value (when not set), we process it as a 0 and that is why I assert based on that. Thank you!

File: internal/service/searchindex/resource_search_index_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (manupedrozo) [reply internal/service/searchindex/resource_search_index_test.go]: I think that the point here is that `var ret int` is not an undefined int value, its always `0`. 
So we can directly check for `"0"` which is what we expect.

The go int zero value will never change, but in the hypothetical case it did, we want to make sure there is no regression in the datasource: we expect a `0`.

File: internal/service/searchindex/resource_search_index_test.go [hunk 10]
```diff
@@ -546,6 +705,31 @@ func checkVector(projectID, indexName, clusterName string) resource.TestCheckFun
...
+		resource.TestCheckResourceAttr(resourceName, "analyzer", "lucene.standard"),
+		resource.TestCheckResourceAttr(resourceName, "search_analyzer", "lucene.standard"),
+		resource.TestCheckResourceAttr(datasourceName, "analyzer", "lucene.standard"),
+		resource.TestCheckResourceAttr(datasourceName, "search_analyzer", "lucene.standard"),
+	}
+
+	if numPartitions != nil {
+		checks = append(checks,
+			resource.TestCheckResourceAttr(resourceName, "num_partitions", fmt.Sprintf("%d", *numPartitions)),
+			resource.TestCheckResourceAttr(datasourceName, "num_partitions", fmt.Sprintf("%d", *numPartitions)),
```
Comment (manupedrozo) [inline internal/service/searchindex/resource_search_index_test.go]:
```suggestion
			resource.TestCheckResourceAttr(resourceName, "num_partitions", strconv.Itoa(*numPartitions)),
			resource.TestCheckResourceAttr(datasourceName, "num_partitions", strconv.Itoa(*numPartitions)),
```

File: internal/service/searchindex/resource_search_index_test.go [hunk 11]
```diff
@@ -533,11 +585,118 @@ func configVector(projectID, indexName, clusterName string) string {
...
+	mappingsDynamic := "true"
+	checks := []resource.TestCheckFunc{
+		resource.TestCheckResourceAttrWith(resourceName, "fields", acc.JSONEquals(fieldsJSON)),
+		resource.TestCheckResourceAttrWith(datasourceName, "fields", acc.JSONEquals(fieldsJSON)),
+	}
+
+	if numPartitions != nil {
+		checks = append(checks,
+			resource.TestCheckResourceAttr(resourceName, "num_partitions", fmt.Sprintf("%d", *numPartitions)),
+			resource.TestCheckResourceAttr(datasourceName, "num_partitions", fmt.Sprintf("%d", *numPartitions)),
```
Comment (manupedrozo) [inline internal/service/searchindex/resource_search_index_test.go]:
```suggestion
			resource.TestCheckResourceAttr(resourceName, "num_partitions", strconv.Itoa(*numPartitions)),
			resource.TestCheckResourceAttr(datasourceName, "num_partitions", strconv.Itoa(*numPartitions)),
```

File: internal/service/searchindex/resource_search_index_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (lantoli) [reply internal/service/searchindex/resource_search_index_test.go]: yep, I think it says that you just do:
```
checks = append(checks,
			resource.TestCheckResourceAttr(resourceName, "num_partitions", "0"),
			resource.TestCheckResourceAttr(datasourceName, "num_partitions", "0"),
		)
```

**Comment by lantoli** [comment]
note that a PR check is failing:
```
=== NAME  TestAccVectorSearchIndex_withNumPartitions
    resource_search_index_test.go:227: Step 1/3 error: Error running apply: exit status 1
        
        Error: error during search deployment creation
        
          with mongodbatlas_search_deployment.test,
          on terraform_plugin_test.tf line 13, in resource "mongodbatlas_search_deployment" "test":
          13: 		resource "mongodbatlas_search_deployment" "test" {
        
        https://cloud-dev.mongodb.com/api/atlas/v2/groups/694199be3d2621ae3a708994/clusters/test-acc-tf-c-858085136270834754/search/deployment
        POST: HTTP 409 Conflict (Error code:
        "ATLAS_SEARCH_DEPLOYMENT_ALREADY_EXISTS") Detail: Search deployment already
        exists for specified cluster. Reason: Conflict. Params: [], BadRequestDetail:
--- FAIL: TestAccVectorSearchIndex_withNumPartitions (2.65s)
```

File: internal/service/searchindex/data_source_search_index.go (hunk 5 above)
Comment (filipcirtog) [reply internal/service/searchindex/data_source_search_index.go:123]: We receive an integer pointer as nil. Since SDKv2 does not natively handle nil values, and our num_partitions field is of type integer, the field will default to an undefined integer value, equal to 0.

**Review by manupedrozo — APPROVED** [review]

File: internal/service/searchindex/resource_search_index_test.go (hunk 11 above)
Comment (filipcirtog) [reply internal/service/searchindex/resource_search_index_test.go]: thank you! changed

File: internal/service/searchindex/resource_search_index_test.go (hunk 10 above)
Comment (filipcirtog) [reply internal/service/searchindex/resource_search_index_test.go]: thank you! changed

File: internal/service/searchindex/resource_search_index_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (filipcirtog) [reply internal/service/searchindex/resource_search_index_test.go]: thank you! changed

File: docs/data-sources/search_indexes.md [hunk 12]
```diff
@@ -52,5 +52,6 @@ data "mongodbatlas_search_indexes" "test" {
 * `type_sets` - Set of type set definitions (when present). Each item includes:
   * `name` - Type set name.
   * `types` - JSON array string describing the types for the set.
+* `num_partitions` - Number of index partitions, returns 0 if not set in the resource
```
[the remark this answers is not in the document]
Comment (oarbusi) [reply docs/data-sources/search_indexes.md]: if default value is 1(it says so in the API docs), why we return 0? seems like an inconsistency in the API?

[the remark this answers is not in the document]
Comment (filipcirtog) [reply docs/data-sources/search_indexes.md]: @oarbusi We receive an integer pointer as nil. Since SDKv2 does not natively handle nil values, and our num_partitions field is of type integer, the field will default to an undefined integer value, equal to 0.

[the remark this answers is not in the document]
Comment (filipcirtog) [reply docs/data-sources/search_indexes.md]: WDYT?

[the remark this answers is not in the document]
Comment (oarbusi) [reply docs/data-sources/search_indexes.md]: it makes sense considering the API bahavior. What I find confusing is the fact that we have a "default is 1" statement but then it "returns 0 when not set", which contradicts the default is 1 statement

[the remark this answers is not in the document]
Comment (filipcirtog) [reply docs/data-sources/search_indexes.md]: In Atlas, if this value is not explicitly set, it defaults to 1. However, this default value is not included in the response

For example:
- If you set the value to 1, the response returns 1
- If you don't set the value, Atlas internally defaults it to 1, but this default value will not appear in the response

[the remark this answers is not in the document]
Comment (oarbusi) [reply docs/data-sources/search_indexes.md]: I still think it can be confusing, but good to resolve on my side

[the remark this answers is not in the document]
Comment (filipcirtog) [reply docs/data-sources/search_indexes.md]: I agree with you! Do you have any suggestions for rephrasing it? Or do you think we should remove the `returns 0` note entirely since that is considered not-set in go? I don't have much experience with such customer-facing documentation.

[the remark this answers is not in the document]
Comment (oarbusi) [reply docs/data-sources/search_indexes.md]: it's a hard one, and I am not sure whether there's an easy solution/or clearer way of explaining this

[the remark this answers is not in the document]
Comment (filipcirtog) [reply docs/data-sources/search_indexes.md]: Update: Removed the comment

Atlas defaults the value to 1 but returns nil unless it's explicitly set and we interpret it as 0. Since both nil and 0 are treated identically in this context, we send nil back to Atlas. As a result, Atlas persists the value as 1 again because that's its default setting unless explicitly overridden.

I initially thought that if we didn't explicitly set the value, we would receive nil, which we interpret as 0. Then, if we attempted the operation again, we would send 0 back, potentially causing issues which turned out as a wrong assumption.

File: internal/service/searchindex/resource_search_index.go [hunk 13]
```diff
@@ -269,13 +273,18 @@ func resourceUpdate(ctx context.Context, d *schema.ResourceData, meta any) diag.
...
+			NumPartitions:  searchRead.LatestDefinition.NumPartitions,
 		},
 	}
 
 	if d.HasChange("analyzer") {
 		searchIndex.Definition.Analyzer = conversion.StringPtr(d.Get("analyzer").(string))
 	}
 
+	if d.HasChange("num_partitions") {
+		searchIndex.Definition.NumPartitions = conversion.IntPtr(d.Get("num_partitions").(int))
```
[the remark this answers is not in the document]
Comment (filipcirtog) [reply internal/service/searchindex/resource_search_index.go:285]: thanks. but we do not do such validation on the client side. we rely on the server (atlas) for that.

**Review by oarbusi — APPROVED** [review]
